src/xunit.analyzers/TestMethodShouldNotHaveReturnType.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Xunit.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class TestMethodShouldNotHaveReturnType : XunitDiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(Descriptors.X1027_TestMethodShouldNotHaveReturnType);
+
+        internal override void AnalyzeCompilation(CompilationStartAnalysisContext compilationStartContext, XunitContext xunitContext)
+        {
+            var compilation = compilationStartContext.Compilation;
+            var taskType = compilation.GetTypeByMetadataName(Constants.Types.SystemThreadingTasksTask);
+
+            compilationStartContext.RegisterSymbolAction(symbolContext =>
+            {
+                var methodSymbol = (IMethodSymbol)symbolContext.Symbol;
+                // NOTE: This analyzer should fire even if the return type is a derived type of Task, such as Task<int>.
+                if (methodSymbol.ReturnsVoid || (taskType != null && methodSymbol.ReturnType == taskType))
+                    return;
+
+                var attributes = methodSymbol.GetAttributes();
+                if (!attributes.ContainsAttributeType(xunitContext.FactAttributeType, exactMatch: true) &&
+                    !attributes.ContainsAttributeType(xunitContext.TheoryAttributeType, exactMatch: true))
```

Comment:
@marcind I'm not very familiar with the xunit extensibility model. I know that people can extend `Fact` and create their own test method attributes, but I'm unsure whether it would be possible to have a fact where there's a good reason for having a non-void/Task return type. Should I keep the current implementation or go with

```cs
// Covers both Fact, Theory, and custom attributes derived from Fact
if (!attributes.ContainsAttributeType(xunitContext.FactAttributeType))
```

Reply:
I guess it's possible someone could be doing something with the result of the test method invocation. But part of me thinks we should just run this on any method decorated with `Fact` or a derived attribute just to see if someone has that scenario. Worst case, the rule can be disabled by the user until we sort it out.